modular_skyrat/modules/primitive_cooking_additions/code/cauldron.dm

```diff
@@ -0,0 +1,339 @@
+	var/static/radial_cook = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_cook")
+
+	/// Radial list options
+	var/static/list/radial_options = list("eject" = radial_eject, "cook" = radial_cook)
+
+/obj/machinery/cauldron/Initialize(mapload)
+	. = ..()
+	register_context()
+	update_appearance(UPDATE_ICON)
+
+/obj/machinery/cauldron/examine(mob/user)
+	. = ..()
+
+	. += span_notice("It can be taken apart with a <b>crowbar</b>.")
+
+	if(!in_range(user, src) && !isobserver(user))
+		. += span_warning("You're too far away to examine [src]'s contents!")
+		return
+
+	if(length(ingredients))
+		. += span_notice("\The [src] contains:")
+		var/list/items_counts = new
+		for(var/i in ingredients)
+			if(isstack(i))
+				var/obj/item/stack/item_stack = i
+				items_counts[item_stack.name] += item_stack.amount
+			else
+				var/atom/movable/single_item = i
+				items_counts[single_item.name]++
+		for(var/item in items_counts)
+			. += span_notice("- [items_counts[item]]x [item].")
+	else
+		. += span_notice("\The [src] is empty.")
+
+/obj/machinery/cauldron/Exited(atom/movable/gone, direction)
+	if(gone in ingredients)
+		ingredients -= gone
+		if(!QDELING(gone) && ingredients.len && isitem(gone))
+			var/obj/item/itemized_ingredient = gone
+			if(!(itemized_ingredient.item_flags & NO_PIXEL_RANDOM_DROP))
+				itemized_ingredient.pixel_x = itemized_ingredient.base_pixel_x + rand(-6, 6)
+				itemized_ingredient.pixel_y = itemized_ingredient.base_pixel_y + rand(-5, 6)
+	return ..()
+
+/obj/machinery/cauldron/on_deconstruction(disassembled)
+	eject()
+	return ..()
+
+/obj/machinery/cauldron/Destroy()
+	QDEL_LIST(ingredients)
+	return ..()
+
+/obj/machinery/cauldron/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	. = ..()
+	if(held_item?.tool_behaviour == TOOL_WRENCH)
+		context[SCREENTIP_CONTEXT_LMB] = "[anchored ? "Unsecure" : "Secure"]"
+		return CONTEXTUAL_SCREENTIP_SET
+
+	if(held_item?.tool_behaviour == TOOL_CROWBAR)
+		context[SCREENTIP_CONTEXT_LMB] = "Deconstruct"
+		return CONTEXTUAL_SCREENTIP_SET
+
+	context[SCREENTIP_CONTEXT_LMB] = "Show menu"
+
+	if(length(ingredients) != 0)
+		context[SCREENTIP_CONTEXT_RMB] = "Start cooking"
+
+	return CONTEXTUAL_SCREENTIP_SET
+
+#define CAULDRON_INGREDIENT_OVERLAY_SIZE 14
+
+/obj/machinery/cauldron/update_overlays()
+	. = ..()
+
+	var/ingredient_count = 0
+
+	for(var/atom/movable/ingredient as anything in ingredients)
+		var/image/ingredient_overlay = image(ingredient, src)
+
+		var/list/icon_dimensions = get_icon_dimensions(ingredient.icon)
+		ingredient_overlay.transform = ingredient_overlay.transform.Scale(
+			CAULDRON_INGREDIENT_OVERLAY_SIZE / icon_dimensions["width"],
+			CAULDRON_INGREDIENT_OVERLAY_SIZE / icon_dimensions["height"],
+		)
+
+		ingredient_overlay.pixel_x = ingredient_shifts_x[(ingredient_count % ingredient_shifts_x.len) + 1]
+		ingredient_overlay.pixel_y = ingredient_shifts_y[(ingredient_count % ingredient_shifts_y.len) + 1]
```

Comment:
pixel_x and pixel_y is for physical location and not to be used on overlays
pixel_w and pixel_z is for aligning overlays